### PR TITLE
Exposes Xaml Application for App authors and Control authors to customize

### PR DESCRIPTION
**Pull request by ocalvo**
Issue: #
Bug 18864996: XamlIslands - Add referenced control's XAMLMetadataProvider to the Interop's metadata provider list
Bug 20507104: XamlApplication needs to call WindowsXamlManager.InitializeForCurrentThread in its constructor
Task 19586623: [OOB] Get GitHub version of WindowsXamlHost to match the version used in tests infrastructure

## PR Type
What kind of change does this PR introduce?

- Bugfix
- Feature

## What is the current behavior?
There is no easy way to customize how and when the Xaml application object is created.
Control authors have no way of adding custom metadata providers.

What is the new behavior?
- We are making XamlApplication a public object, and exposing as a protected static property: MetadataProviders for control authors, this resolves bug 18864996.
Sample code:
        public class MyControl : Toolkit.Wpf.UI.XamlHost.WindowsXamlHostBase
        {
            static MyControl()
            {
                if (Application != null)
                {
                    Application.MetadataProviders.Add(new MyMetadataProvider());
                }
                else
                {
                    throw new InvalidOperationException($"{typeof(MyControl).Name} only supported when Windows.UI.Xaml.Application.Current is {typeof(Win32.UI.XamlHost.XamlApplication).FullName}");
                }
            }
            public MyControl()
            {
            }
        }

- We are making the constructor of XamlApplication to call WindowsXamlManager.InitializeForCurrentThread and Disposable, this fixes bug 20507104 and allows customers for customization of the XamlApplication at startup:
Sample code:
        public static void Main()
        {
            using (var xamlApp = new Win32.UI.XamlHost.XamlApplication()
            {
                Resources = new Windows.UI.Xaml.ResourceDictionary()
                    {
                        { "MyResourceKey", "MyValue" },
                    },
                MetadataProviders = {
                    new MyMetadataProvider(),
                },
            })
            {
                var app = new Microsoft.Toolkit.Sample.Wpf.App.App();
                app.InitializeComponent();
                app.Run();
            }
        }
        public static void Main()
        {
            using (var xamlApp = new Win32.UI.XamlHost.XamlApplication())
            {
                var appOwnedWindowsXamlManager = xamlApp.WindowsXamlManager;
                var app = new Microsoft.Toolkit.Sample.Wpf.App.App();
                app.InitializeComponent();
                app.Run();
            }
        }
- At the same time, we are NOT breaking any compatibility with previous code that references the toolkit as we are not introducing any breaking change.
a.	We still support WPF and WinForms apps that never reference XamlApplication.
b.	We still support any Win32 app that creates its own custom application object.
i.	In this case they will have to still work themselves the details on how to hook the proper metadata providers.
- We are introducing a new assembly:  Microsoft.Toolkit.Win32.UI.XamlHost.dll that contains XamlApplication and other common code between WPF and WinForms.
a.	This is an important change for Apps that use WPF and WinForms at the same time, as before this change there was 2 different types for XamlApplication (one for WPF and one for WinForms)
b.	This made it impossible to use toolkits controls from WPF or WinForms in the same win32 app.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: 
- [ ] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [ ] Contains **NO** breaking changes

## Other information
No breaking changes are introduced.

### CODE REVIEW

**Comment by devsko** [comment]
Thats is almost what I was looking for. Solves one and a half of my problems.
 
About `XamlApplication.MetadataProviders`:
- it would be great if the default `MetadataProviderDiscovery` would be optional now where I can set the providers manually. I have hundreds of assemblies in my self contained application. Especially while debugging it takes minutes to load them all.
- it seems there is a timing problem how the providers are added to the `IList`. The `XamlApplication.ctor` calls `WindowsXamlManager.InitializeForCurrentThread` which already calls `GetXmlnsDefinitions` which initializes `MetadataProviders`. The code that adds `MyMetadataProvider` runs later. (After it run, there are 2 instances of `MyMetadataProvider` in the list because one was discovered and created by `MetadataProviderDiscovery`)

File: Microsoft.Toolkit.Sample.Wpf.App/Program.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                using (var xamlApp = new Win32.UI.XamlHost.XamlApplication()
+                {
+                    Resources = new Windows.UI.Xaml.ResourceDictionary()
+                        {
+                            { "MyResourceKey", "MyValue" },
+                        },
+                    MetadataProviders = {
+                        new MyMetadataProvider(),
+                    },
+                })
```
Comment (devsko) [inline Microsoft.Toolkit.Sample.Wpf.App/Program.cs]: Here MyMetadataProvider gets added too late. The .ctor already uses MetadataProviders

File: Microsoft.Toolkit.Win32.UI.XamlHost/XamlApplication.Disposable.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                    return new XamlApplication();
+                }
+                catch
+                {
+                    _application = Windows.UI.Xaml.Application.Current;
+                }
+            }
+
+            var xamlApplication = _application as XamlApplication;
+            if (xamlApplication != null && xamlApplication.IsDisposed)
```
Comment (devsko) [inline Microsoft.Toolkit.Win32.UI.XamlHost/XamlApplication.Disposable.cs]:
```suggestion
            if (_application is XamlApplication xamlApplication && xamlApplication.IsDisposed)
```

**Comment by ocalvo** [comment]
In the latest iteration the default constructor of XamlApplication will not add MetadataProviderDiscovery/
The timing problem  you describe is by design, when initializing Xaml, it itself queries the metadata providers. But this is not a problem.


---
In reply to: [466310766](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/82#issuecomment-466310766) [](ancestors = 466310766)

File: Microsoft.Toolkit.Sample.Wpf.App/Program.cs (hunk 1 above)
Comment (ocalvo) [reply Microsoft.Toolkit.Sample.Wpf.App/Program.cs]: Resolved now, you should see only one instance.

---
In reply to: [259245254](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/82#discussion_r259245254) [](ancestors = 259245254)

**Comment by ocalvo** [comment]
Supperceded by https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/85

**Comment by ocalvo** [comment]
Supperceded by https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/85

**Pull request closed without merging**